src/compiler/checker.ts

```diff
@@ -13523,6 +13535,22 @@ namespace ts {
                 return type;
             }
 
+            function narrowTypeByConstructor(type: Type, expr: Expression, operator: SyntaxKind, assumeTrue: boolean): Type {
+                if (!assumeTrue || isExclamationEqualsOrExclamationEqualsEqualsKind(operator)) return type;
+
+                const rightType = getTypeOfExpression(expr);
+                if (!isTypeSubtypeOf(rightType, globalFunctionType)) return type;
+
+                const prototypeProperty = getPropertyOfType(rightType, "prototype" as __String);
+                if (!prototypeProperty) return type;
+
+                const prototypePropertyType = getTypeOfSymbol(prototypeProperty);
+                const targetType = !isTypeAny(prototypePropertyType) ? prototypePropertyType : undefined;
+                if (!targetType || isTypeAny(type) && (targetType === globalObjectType || targetType === globalFunctionType)) return type;
+
+                return getNarrowedType(type, targetType, assumeTrue, areTypesComparable);

```

Comment:
I don't think you want `areTypesComparable` because this will work for types which don't explicitly derive from a given constructor.

I think you potentially want a helper function that uses `isTypeDerivedFrom` on the wrapper object type for any primitive types.

Reply:
> will work for types which don't explicitly derive from a given constructor

could you give some example? 😢